dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,758 @@
+        checkAttributeType(name, attribute_name, dict_struct.attributes[ind].underlying_type, AttributeUnderlyingType::ut##TYPE); \
+\
+        getItemsImpl<TYPE, TYPE>( \
+            ind, key_columns, [&](const size_t row, const auto value) { out[row] = value; }, [&](const size_t) { return def; }); \
+    }
+    DECLARE(UInt8)
+    DECLARE(UInt16)
+    DECLARE(UInt32)
+    DECLARE(UInt64)
+    DECLARE(UInt128)
+    DECLARE(Int8)
+    DECLARE(Int16)
+    DECLARE(Int32)
+    DECLARE(Int64)
+    DECLARE(Float32)
+    DECLARE(Float64)
+    DECLARE(Decimal32)
+    DECLARE(Decimal64)
+    DECLARE(Decimal128)
+#undef DECLARE
+
+void IPolygonDictionary::getString(
+        const std::string & attribute_name,
+        const Columns & key_columns,
+        const DataTypes &,
+        const String & def,
+        ColumnString * const out) const
+{
+    const auto ind = getAttributeIndex(attribute_name);
+    checkAttributeType(name, attribute_name, dict_struct.attributes[ind].underlying_type, AttributeUnderlyingType::utString);
+
+    getItemsImpl<String, StringRef>(
+            ind,
+            key_columns,
+            [&](const size_t, const StringRef value) { out->insertData(value.data, value.size); },
+            [&](const size_t) { return StringRef{def}; });
+}
+
+template <typename AttributeType, typename OutputType, typename ValueSetter, typename DefaultGetter>
+void IPolygonDictionary::getItemsImpl(
+        size_t attribute_ind, const Columns & key_columns, ValueSetter && set_value, DefaultGetter && get_default) const
+{
+    const auto points = extractPoints(key_columns);
+
+    using ColVecType = std::conditional_t<IsDecimalNumber<AttributeType>, ColumnDecimal<AttributeType>, ColumnVector<AttributeType>>;
+    using ColType = std::conditional_t<std::is_same<AttributeType, String>::value, ColumnString, ColVecType>;
+    const auto column = typeid_cast<const ColType *>(attributes[attribute_ind].get());
+    if (!column)
+        throw Exception{"An attribute should be a column of its type", ErrorCodes::BAD_ARGUMENTS};
+    for (const auto i : ext::range(0, points.size()))
+    {
+        size_t id = 0;
+        const auto found = find(points[i], id);
+        id = ids[id];
+        if (!found)
+        {
+            set_value(i, static_cast<OutputType>(get_default(i)));
+            continue;
+        }
+        if constexpr (std::is_same<AttributeType, String>::value)
+            set_value(i, static_cast<OutputType>(column->getDataAt(id)));
+        else
+            set_value(i, static_cast<OutputType>(column->getElement(id)));
+    }
+
+    query_count.fetch_add(points.size(), std::memory_order_relaxed);
+}
+
+namespace
+{
+
+inline void makeDifferences(IColumn::Offsets & values)
+{
+    for (size_t i = 1; i < values.size(); ++i)
+        values[i] -= values[i - 1];
+}
+
+struct Offset
+{
+    Offset() = default;
+    IColumn::Offsets ring_sizes, polygon_sizes, multi_polygon_sizes;
```

Comment:
Do not declare vars in single line. It's against out code style.

Reply:
Okay, I'll keep that in mind.